browser/onboarding/onboarding_tab_helper.cc

```diff
@@ -0,0 +1,113 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/onboarding/onboarding_tab_helper.h"
+
+#include <string>
+#include <vector>
+
+#include "base/strings/string_number_conversions.h"
+#include "brave/browser/onboarding/domain_map.h"
+#include "brave/browser/onboarding/pref_names.h"
+#include "brave/browser/ui/views/frame/brave_browser_view.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/first_run/first_run.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "components/grit/brave_components_strings.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace onboarding {
+
+OnboardingTabHelper::~OnboardingTabHelper() = default;
+
+OnboardingTabHelper::OnboardingTabHelper(content::WebContents* web_contents)
+    : content::WebContentsObserver(web_contents),
+      content::WebContentsUserData<OnboardingTabHelper>(*web_contents) {}
+
+void OnboardingTabHelper::DidStopLoading() {
+  Browser* browser = chrome::FindBrowserWithWebContents(web_contents());
+  auto* shields_data_controller =
+      brave_shields::BraveShieldsDataController::FromWebContents(
+          web_contents());
+  auto* active_tab_web_contents =
+      browser->tab_strip_model()->GetActiveWebContents();
+
+  if (web_contents() != active_tab_web_contents)
+    return;
+
+  if (shields_data_controller &&
+      shields_data_controller->GetBraveShieldsEnabled() &&
+      shields_data_controller->GetTotalBlockedCount() > 1 &&
+      CanHighlightBraveShields()) {
+    ShowBraveHelpBubbleView();
+  }
+}
+
+void OnboardingTabHelper::ShowBraveHelpBubbleView() {
+  Browser* browser = chrome::FindBrowserWithWebContents(web_contents());
+  BraveBrowserView* browser_view = static_cast<BraveBrowserView*>(
+      BraveBrowserView::GetBrowserViewForBrowser(browser));
+  browser_view->ShowBraveHelpBubbleView(GetTextForOnboardingShieldsBubble());
+
+  g_browser_process->local_state()->SetBoolean(
+      prefs::kOnboardingIsShieldsHighlighted, true);
+}
+
+bool OnboardingTabHelper::CanHighlightBraveShields() {
+  base::ScopedAllowBlockingForTesting allow_blocking;
+  base::Time time_first_run = first_run::GetFirstRunSentinelCreationTime();
+  base::Time time_now = base::Time::Now();
+  base::Time time_7_days_ago = time_now - base::Days(7);
+  bool has_browser_been_run_recently = (time_first_run > time_7_days_ago);
```

Comment:
Our rollout strategy for this is to only show highlight bubble to users that has installed the browser in the previous 7 days or if its a first run, after this PR merges. 

I'm going to modify the condition here to include if the browser ran for the first time as well and this would work if the browser is started with `--force-first-run` for testing purpose.

```c++ 
if ((first_run::IsChromeFirstRun() || has_browser_been_run_recently) && !is_shields_highlighted) {}
```

Reply:
I think it makes sense to check is_shields_highlighted first and then sentinel creation time as it's a IO task